api/src/main/java/com/datastrato/gravitino/rel/partitions/Partitions.java

```diff
@@ -0,0 +1,234 @@
+          && Objects.equals(properties, that.properties);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(name, upper, lower, properties);
+    }
+  }
+
+  /** Represents a result of list partitioning. */
+  public static class ListPartition implements Partition {
+    private final String name;
+    private final Literal<?>[][] lists;
+
+    private final Map<String, String> properties;
+
+    private ListPartition(String name, Literal<?>[][] lists, Map<String, String> properties) {
+      this.name = name;
+      this.properties = properties;
+      this.lists = lists;
+    }
+
+    /** @return The values of the list partition. */
+    public Literal<?>[][] lists() {
+      return lists;
+    }
+
+    @Override
+    public String name() {
+      return name;
+    }
+
+    @Override
+    public Map<String, String> properties() {
+      return properties;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      ListPartition that = (ListPartition) o;
+      return Objects.equals(name, that.name)
+          && Arrays.deepEquals(lists, that.lists)
+          && Objects.equals(properties, that.properties);
+    }
+
+    @Override
+    public int hashCode() {
+      int result = Objects.hash(name, properties);
+      result = 31 * result + Arrays.deepHashCode(lists);
+      return result;
+    }
+  }
+
+  /** Represents a result of identity partitioning. */
+  public static class IdentityPartition implements Partition {
+    private final String name;
+    private final String[][] fieldNames;
+    private final Literal<?>[] values;
+    private final Map<String, String> properties;
+
+    private IdentityPartition(
+        String name, String[][] fieldNames, Literal<?>[] values, Map<String, String> properties) {
+      this.name = name;
+      this.fieldNames = fieldNames;
+      this.values = values;
+      this.properties = properties;
+    }
+
+    /** @return The field names of the identity partition. */
+    public String[][] fieldNames() {
+      return fieldNames;
+    }
+
+    /** @return The values of the identity partition. */
+    public Literal<?>[] values() {
+      return values;
+    }
+
+    @Override
+    public String name() {
+      return name;
+    }
+
+    @Override
+    public Map<String, String> properties() {
+      return properties;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      IdentityPartition that = (IdentityPartition) o;
+      return Objects.equals(name, that.name)
+          && Arrays.deepEquals(fieldNames, that.fieldNames)
+          && Arrays.equals(values, that.values)
+          && Objects.equals(properties, that.properties);
```

Comment:
Do we need to compare the properties for a partition?

Reply:
Good question, I think it's necessary. 

Assuming a scenario where the user gets partition A' after inserting records into partition A using the partition name. In this case, the property rowCount value may change and it is important to note that partition A' should not be equal to partition A.

Reply:
OK.